### PR TITLE
feat(lambda,event-bridge): lambda event sources

### LINKED ISSUE #44: Event Sources
Add auto event bus rule generation from integrations with the ability.

Some AWS services like Lambda, Step Functions, and more automatically generate Rules on the default bus.

Add the ability to automatically create these rules and then send to all Functionless Event Bus target. They should support type safe event types based on the event.

```ts
myFunction.onSucccess().send(myMachine, event => event.detail.result);
myMachine.onStarted().send(...)
```

Where `send` is the outcome of #37 's final target design.

### PR DESCRIPTION
**Pull request by thantos**
Closes #44 

Depends on #161 - This change made major changes to EventBus, stacking these changes to reduce churn.

* [x] Lambda
* [x] Step Functions - Already done
❌ Api Gateway
❌ App Sync
❌ Dynamo

 ## Lambda

Unlike Step Functions, Lambda does not emit events by default. Lambda has a concept of a `destination` or `async invocation`. Each lambda, version, and alias can only have a single onSuccess and onFailure destination. Destinations can be EventBuses, Sns topics, Sqs queues, or other lambda functions.

```ts
const func = new Function(stack, 'func', {
   onSuccess: bus, // IEventBus | IFunction | aws_lambda.IDestination
   onFailure: bus // IEventBus | IFunction | aws_lambda.IDestination
});

// fail when onSuccess is not set and not a bus?
func.onSuccess(bus).map().pipe();
func.onFailure(bus).map().pipe();
```

### CODE REVIEW

File: src/function.ts [hunk 1]
```diff
@@ -140,6 +208,56 @@ abstract class FunctionBase<P, O>
...
+  ): Rule<AsyncResponseFailureEvent<P>> {
+    return new PredicateRuleBase<AsyncResponseFailureEvent<P>>(
+      bus.bus,
+      id,
+      bus,
+      /**
+       * when(event => event.source === "lambda"
+       *    && event["detail-type"] === "Lambda Function Invocation Result - Failure"
+       *      && event.resources.includes(this.resource.functionArn))
+       */
```
[the remark this answers is not in the document]
Comment (thantos) [reply src/function.ts:338]: yes :(

File: src/event-bridge/rule.ts [hunk 2]
```diff
@@ -138,8 +138,8 @@ export interface IRule<T extends Event> {
    */
   pipe(props: LambdaTargetProps<T>): void;
   pipe(func: IFunction<T, any>): void;
-  pipe(bus: IEventBus<T>): void;
-  pipe(props: EventBusTargetProps<T>): void;
+  pipe(bus: IEventBus<any>): void;
+  pipe(props: EventBusTargetProps<any>): void;
```
Comment (sam-goodwin) [inline src/event-bridge/rule.ts:142]: Why `any`? Wouldn't that break type safety?

File: src/function.ts [hunk 3]
```diff
@@ -52,8 +83,112 @@ export interface IFunction<P, O> {
...
+   * and that destination must match the bus provided here.
+   *
+   * @see https://docs.aws.amazon.com/lambda/latest/dg/invocation-async.html#invocation-async-destinations
+   *
+   * The rule returned will contain the logic:
+   *
+   * when(id, event => event.source === "lambda"
+   *    && event["detail-type"] === "Lambda Function Invocation Result - Success"
+   *      && event.resources.includes(this.resource.functionArn))
+   */
```
Comment (sam-goodwin) [inline src/function.ts:101]: Surround with triple backticks block. Tsdoc produces markdown

File: src/function.ts [hunk 4]
```diff
@@ -140,6 +275,78 @@ abstract class FunctionBase<P, O>
...
+
+  configureAsyncInvoke(config: EventInvokeConfigOptions<P, O>): void {
+    this.resource.configureAsyncInvoke({
+      ...config,
+      onSuccess: FunctionBase.normalizeAsyncDestination<P, O>(config.onSuccess),
+      onFailure: FunctionBase.normalizeAsyncDestination<P, O>(config.onFailure),
+    });
+  }
+
+  onSuccess(
```
Comment (sam-goodwin) [inline src/function.ts:301]: Does this inherit docs?

File: src/function.ts [hunk 5]
```diff
@@ -164,6 +374,9 @@ export interface FunctionProps
   clientConfigRetriever?: (
     clientName: ClientName | string
   ) => Omit<AWS.Lambda.ClientConfiguration, keyof AWS.Lambda.ClientApiVersions>;
+
+  onSuccess?: FunctionAsyncOnSuccessDestination<P, O>;
+  onFailure?: FunctionAsyncOnFailureDestination<P>;
```
Comment (sam-goodwin) [inline src/function.ts:379]: Missing docs

File: src/function.ts [hunk 6]
```diff
@@ -140,6 +275,78 @@ abstract class FunctionBase<P, O>
       ResultSelector: "$.Payload",
     };
   }
+
+  protected static normalizeAsyncDestination<P, O>(
```
Comment (sam-goodwin) [inline src/function.ts:279]: Why protected?

File: src/function.ts [hunk 7]
```diff
@@ -140,6 +275,78 @@ abstract class FunctionBase<P, O>
...
+    return destination === undefined
+      ? undefined
+      : isEventBus<any>(destination)
+      ? new EventBridgeDestination(destination.bus)
+      : isFunction(destination)
+      ? new LambdaDestination(destination.resource)
+      : destination;
+  }
+
+  configureAsyncInvoke(config: EventInvokeConfigOptions<P, O>): void {
```
Comment (sam-goodwin) [inline src/function.ts:293]: Should this be public or private? I didn't see it in the interface so private?

File: src/function.ts (hunk 4 above)
Comment (sam-goodwin) [reply src/function.ts:301]: No was just curious if it was happpening implicitly?

Comment (thantos) [reply src/function.ts:301]: Yeah, it inherits from the interface.

File: src/function.ts (hunk 6 above)
Comment (thantos) [reply src/function.ts:279]: This is an abstract base class, this method is used by the base and child class. And we don't need to expose this.

File: src/function.ts (hunk 7 above)
Comment (thantos) [reply src/function.ts:293]: Its in the interface, made it public

image

Comment (thantos) [reply src/function.ts:293]: Added an es-lint rule to error on these cases to make your life easier.

**Comment by thantos** [comment]
Note: prettier is broken due to the use of typescript 4.7

Waiting on: https://github.com/prettier/prettier/issues/12640

File: src/function.ts [hunk 8]
```diff
@@ -44,17 +57,154 @@ export type FunctionClosure<P, O> = (
...
 
-export interface IFunction<P, O> {
+/**
+ * Returns the payload type on the {@link IFunction}.
+ */
+export type FunctionPayload<F extends IFunction<any, any>> = [F] extends [IFunction<infer P, any>] ? P : never;
+/**
+ * Returns the output type on the {@link IFunction}.
+ */
+ export type FunctionOutput<F extends IFunction<any, any>> = [F] extends [IFunction<any, infer O>] ? O : never;
```
Comment (sam-goodwin) [inline src/function.ts:67]: nit: indentation

Comment (sam-goodwin) [reply src/function.ts:67]: why the `[F] extends `? What's the purpose of packing it a 1-tuple?

File: src/function.ts [hunk 9]
```diff
@@ -44,17 +57,154 @@ export type FunctionClosure<P, O> = (
...
+   *
+   * The rule returned will contain the logic:
+   *
+   * ```ts
+   * when(id, event => event.source === "lambda"
+   *    && event["detail-type"] === "Lambda Function Invocation Result - Success"
+   *      && event.resources.includes(this.resource.functionArn))
+   * ```
+   */
+  onSuccess(
```
Comment (sam-goodwin) [inline src/function.ts:115]: nit: looks like eslint missed this missing `public`

File: src/function.ts [hunk 10]
```diff
@@ -44,17 +57,154 @@ export type FunctionClosure<P, O> = (
...
   (...args: Parameters<ConditionalFunction<P, O>>): ReturnType<
     ConditionalFunction<P, O>
   >;
+
+  /**
+   * Event Source for the {@link Function} onSuccess async invocation destination.
+   *
+   * For Lambda, the onSuccess destination must first be configured via either
+   * the {@link Function} constructor or by using {@link IFunction.configureAsyncInvoke}
+   * and that destination must match the bus provided here.
```
Comment (sam-goodwin) [inline src/function.ts:103]: I'm not sure what that means and how to take action on it.

File: src/function.ts (hunk 8 above)
Comment (thantos) [reply src/function.ts:67]: `[F] extends []` changes the way that union types are interpreted.

https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types

With `[F]`

image

Without

image

Which in the case its being used in... if the types are not blown out (`Function<A | B>`) then the `isFunction` does not narrow the types correctly. 

```ts
if(isFunction(destination)) {
    destination // is IFunction<AsyncResponseSuccess<P, O> | AsyncResponseFailure<P>, any>
} else {
     destination // is IDestination | IFunction<AsyncResponseSuccess<P, O> | AsyncResponseFailure<P>, any>
}
```

image

But with... we get nice type narrowing.

```ts
if(isFunction(destination)) {
    destination // is IFunction<AsyncResponseSuccess<P, O>, any> | IFunction<AsyncResponseFailure<P>, any>
} else {
     destination // is IDestination
}
```

image

File: src/function.ts (hunk 9 above)
Comment (thantos) [reply src/function.ts:115]: Thats the interface :-)

File: src/function.ts (hunk 8 above)
Comment (sam-goodwin) [reply src/function.ts]: Awesome analysis. Thanks!

File: src/function.ts (hunk 9 above)
Comment (sam-goodwin) [reply src/function.ts:150]: Ahhhh my bad. Thumbs down feels bad lol

File: src/event-bridge/event-bus.ts [hunk 11]
```diff
@@ -40,7 +40,12 @@ export const isEventBus = <E extends Event>(v: any): v is IEventBus<E> => {
   );
 };
 
-export interface IEventBusFilterable<E extends Event> {
+/**
+ * Returns the {@link Event} type on the {@link EventBus}.
+ */
+export type EventBusEvent<B extends IEventBus<any>> = [B] extends [IEventBus<infer E>] ? E : never;
```
Comment (sam-goodwin) [inline src/event-bridge/event-bus.ts:46]: Should this be called GetEventBusType? Or else it could be confused with the Event type. The verb helps with understanding

File: src/event-bridge/event-bus.ts [hunk 12]
```diff
@@ -115,16 +120,16 @@ export interface IEventBusFilterable<E extends Event> {
   ): Rule<E, O>;
 }
 
-export interface IEventBus<E extends Event = Event>
+export interface IEventBus<in E extends Event = Event, O extends E = E>
```
Comment (sam-goodwin) [inline src/event-bridge/event-bus.ts:123]: Might want to document the type parameters. Not immediately obvious what E and O mean

File: src/event-bridge/event-bus.ts [hunk 13]
```diff
@@ -339,7 +339,7 @@ abstract class EventBusBase<E extends Event> implements IEventBus<E> {
   }
 
   public readonly eventBus = makeEventBusIntegration<
-    E,
+    EventBusPutEventInput<O>,
```
Comment (sam-goodwin) [inline src/event-bridge/event-bus.ts:342]: A bit verbose.

File: src/function.ts [hunk 14]
```diff
@@ -64,13 +116,153 @@ export interface IFunction<P, O>
...
+  /**
+   * Event Source for the {@link Function} onSuccess async invocation destination.
+   *
+   * For Lambda, the onSuccess destination is not enabled by default.
+   * It must first be configured via either the {@link Function} constructor
+   * or by using {@link IFunction.configureAsyncInvoke} and that destination must match the bus provided here.
+   *
+   * ```ts
+   * const bus = new EventBus(stack, 'bus');
+   * new Function(stack, 'func', { onSuccess: bus }, async () => {});
```
Comment (sam-goodwin) [inline src/function.ts:128]: Cool!

File: src/function.ts [hunk 15]
```diff
@@ -64,13 +116,153 @@ export interface IFunction<P, O>
...
+   * ```
+   *
+   * or
+   *
+   * ```ts
+   * const bus = new EventBus(stack, 'bus');
+   * const func = new Function(stack, 'func', async () => {});
+   * // if onSuccess or onFailure is already set, this will fail.
+   * func.configureAsyncInvoke({ onSuccess: bus });
```
Comment (sam-goodwin) [inline src/function.ts:137]: Hmm configureAsyncInvoke isn't telling me exactly what it means. This is enabling async invoke and routing events to the bus? Maybe `enableAsyncInvoke`?

**Review by sam-goodwin — APPROVED** [review]

File: src/event-bridge/event-bus.ts (hunk 12 above)
Comment (sam-goodwin) [reply src/event-bridge/event-bus.ts:123]: In standard libraries, I usually prefer to use words for type parameters instead of symbols.

File: src/function.ts (hunk 15 above)
Comment (thantos) [reply src/function.ts:137]: This is the name from the CDK, which I agree, is also confusing..will take another pass.

**Comment by sam-goodwin** [comment]
Should the title of the PR be more granular? Would someone reading the CHANGELOG know what this change is about based purely on the title?

**Comment by thantos** [comment]
> Should the title of the PR be more granular? Would someone reading the CHANGELOG know what this change is about based purely on the title?

Ha... the change started with higher ambitions and then I realized that no one else but lambda and step function publish events!

Will update.

File: src/event-bridge/event-bus.ts (hunk 12 above)
Comment (thantos) [reply src/event-bridge/event-bus.ts:123]: Thanks, that was a helpful exercise. 

```ts
/**
 * @typeParam Evnt - the union type of events that this EventBus can accept.
 *                   `Evnt` is the covariant version of `OutEnvt` in that
 *                   the bus will accept any of `Evnt` while the EventBus can
 *                   emit any of `OutEvnt`.
 * @typeParam OutEvnt - the union type of events that this EventBus will emit through rules.
 *                      `OutEvnt` is the contravariant version of `Envt` in that
 *                      the bus will emit any of `OutEvnt` while the EventBus can
 *                      can accept any of `Evnt`. This type parameter should be left
 *                      empty to be inferred.
 */
```

File: src/event-bridge/rule.ts [hunk 16]
```diff
@@ -21,11 +21,15 @@ import type { Event } from "./types";
...
+export type RulePredicateFunction<Evnt, OutEvnt extends Evnt = Evnt> =
+  | ((event: Evnt) => event is OutEvnt)
+  | ((event: Evnt) => boolean);
 
-export interface IRule<T extends Event> {
+/**
+ * @typeParam - Evnt - The original event type from the {@link EventBus}.
+ * @typeParam - OutEvnt - The narrowed event type after the predicate is applied.
+ */
+export interface IRule<in Evnt extends Event, out OutEvnt extends Evnt> {
```
Comment (sam-goodwin) [inline src/event-bridge/rule.ts:32]: Why do we need two type parameters?

File: src/event-bridge/event-bus.ts [hunk 17]
```diff
@@ -103,28 +108,42 @@ export interface IEventBusFilterable<E extends Event> {
...
    */
-  when<O extends E>(
+  when<InEvnt extends OutEvnt, NewEvnt extends InEvnt>(
     id: string,
-    predicate: RulePredicateFunction<E, O>
-  ): Rule<E, O>;
-  when<O extends E>(
+    predicate: RulePredicateFunction<InEvnt, NewEvnt>
+  ): Rule<InEvnt, NewEvnt>;
+  when<InEvnt extends OutEvnt, NewEvnt extends InEvnt>(
```
Comment (sam-goodwin) [inline src/event-bridge/event-bus.ts:119]: These two parameters seem overly complex and redundant. Why did you add them?

File: src/event-bridge/event-bus.ts [hunk 18]
```diff
@@ -103,28 +108,42 @@ export interface IEventBusFilterable<E extends Event> {
...
-  ): Rule<E, O>;
+    predicate: RulePredicateFunction<InEvnt, NewEvnt>
+  ): Rule<InEvnt, NewEvnt>;
 }
 
-export interface IEventBus<E extends Event = Event>
-  extends IEventBusFilterable<E>,
+/**
+ * @typeParam Evnt - the union type of events that this EventBus can accept.
+ *                   `Evnt` is the covariant version of `OutEvnt` in that
```
Comment (sam-goodwin) [inline src/event-bridge/event-bus.ts:128]: `Evnt` is contra-variant, not co-variant. `out` is for covariance

File: src/event-bridge/event-bus.ts [hunk 19]
```diff
@@ -103,28 +108,42 @@ export interface IEventBusFilterable<E extends Event> {
...
+ *                   the bus will accept any of `Evnt` while the EventBus can
+ *                   emit any of `OutEvnt`.
+ * @typeParam OutEvnt - the union type of events that this EventBus will emit through rules.
+ *                      `OutEvnt` is the contravariant version of `Evnt` in that
+ *                      the bus will emit any of `OutEvnt` while the EventBus can
+ *                      can accept any of `Evnt`. This type parameter should be left
+ *                      empty to be inferred. ex: `EventBus<Event<Detail1> | Event<Detail2>>`.
+ */
+export interface IEventBus<in Evnt extends Event = Event, OutEvnt extends Evnt = Evnt>
```
Comment (sam-goodwin) [inline src/event-bridge/event-bus.ts:137]: This doesn't seem right. I think we can remove the `OutEvnt`

File: src/event-bridge/rule.ts (hunk 16 above)
Comment (thantos) [reply src/event-bridge/rule.ts:32]: `Evnt` is the event in. The input to the predicate on the rule.

`OutEvnt` is the narrowed type as the input to `pipe` or `map` or `when`.

Essentially splitting and showing the two sides of the predicate.

Comment (sam-goodwin) [reply src/event-bridge/rule.ts:32]: Why do you need `Evnt`?

File: src/event-bridge/event-bus.ts (hunk 17 above)
Comment (thantos) [reply src/event-bridge/event-bus.ts:119]: Need to invert the `in`Evnt` to `out NewEvnt` and `out OutEvnt` to `in InEvnt` for the when method.

File: src/event-bridge/rule.ts (hunk 16 above)
Comment (thantos) [reply src/event-bridge/rule.ts:32]: Its being used on `integration: IntegrationWithEventBus<Evnt, Props>,`

But that doesn't seem right... will look.

**Review by sam-goodwin — APPROVED** [review]
We can modify the types in a subsequent change. Ship it

**Comment by thantos** [comment]
> We can modify the types in a subsequent change. Ship it

Removed the unessecary `in Evnt` from `IRule` and `RuleBase` which cleaned up a bit.

Also fixed the covariance/contravariant naming issues and think I understand them better now :-)

https://dmitripavlutin.com/typescript-covariance-contravariance/

**Comment by thantos** [comment]
> Error: Cannot find module '/opt/build/repo/scripts/compile-error-code-page'

I don't understand what changed
